### PR TITLE
[6X] Remove `getaddrinfo` in `SendDummyPacket()` to address malloc deadlock

### PR DESCRIPTION
**Pull request by Chibin**
`SendDummyPacket` eventually calls `getaddrinfo` (which is a reentrant), however, `getaddrinfo` is not an async-signal-safe function. `getaddrinfo` internally calls `malloc`, which is strongly advised to not do within a signal handler as it may cause deadlocks. Cache the accepted socket information for the listener, so that it can be reused in `SendDummyPacket()`.

The purpose of `SendDummyPacket` is to exit more quickly; it circumvents the polling that happens, which eventually times out after 250ms.

Without `SendDummyPacket()`, there will be multiple test failures since some tests expects the backend connection to terminate almost immediately.

To view all the async-signal-safe functions, please view the signal-safety(7) — Linux manual page.

====================

Accommodate for AF_INET6 and AF_INET when doing a motion layer IPC teardown

Previously on commit 70306db18e2, we removed pg_getaddrinfo_all for signal handlers. However, in doing so, the capability of supporting both AF_INET6 and AF_INET was lost; this responsibility must now be handled by us. The commit mentioned above fixed the issue for AF_INET (IPv4), but not for AF_INET6 (IPv6).

This commit also addresses the situation for both AF_INET and AF_INET6.

====================

This commit backports: https://github.com/greenplum-db/gpdb/commit/70306db18e2bfac4969bf327f5ada45bf33f57a1 and
https://github.com/greenplum-db/gpdb/commit/1ee34a424548e759e82c93cf9b4fe1411465e0e9.
Conflicts resolved:
- setupUDPListeningSocket is not one to one with 7X, so resolve conflict
  on the location where `listenerSockaddr` is `memcpy`ed.
- resolve conflicts due to difference of from the modernized version and
  refactor from https://github.com/greenplum-db/gpdb/commit/50748b756758db19229c47911e4c29cbe5ba937b

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by Aegeaner** [comment]
Yeah there are also reports from the community said [glibc's `getaddrinfo()` is also not **_thread-safe_**](https://github.com/golang/go/issues/63567):

> The C setenv function is documented as not being thread-safe ("MT-Unsafe const:env"). Currently, os.Setenv calls C setenv, so it is violating the specification. When this happens, C calls to getenv, either explicitly or in the C library itself (e.g. getaddrinfo, mktime, others), can crash.

See: https://rachelbythebay.com/w/2023/10/16/env/

File: src/backend/cdb/motion/ic_udpifc.c [hunk 1]
```diff
@@ -1282,17 +1291,18 @@ setupUDPListeningSocket(int *listenerSocketFd, uint16 *listenerPort, int *txFami
...
+			continue;
 		}
 
-		if (fd >= 0)
-		{
-			closesocket(fd);
-			fd = -1;
-		}
+		*txFamily = rp->ai_family;
+		break;				/* Success */
```
Comment (soumyadeep2007) [inline src/backend/cdb/motion/ic_udpifc.c:1305]: Ah these changes are from https://github.com/greenplum-db/gpdb/commit/50748b756758db19229c47911e4c29cbe5ba937b

Can we omit these from this patch, to keep this one focused?

File: src/backend/cdb/motion/ic_udpifc.c [hunk 2]
```diff
@@ -1322,6 +1332,13 @@ setupUDPListeningSocket(int *listenerSocketFd, uint16 *listenerPort, int *txFami
 	else
 		*listenerPort = ntohs(((struct sockaddr_in *) &our_addr)->sin_port);
 
+	/*
+	 * cache the successful sockaddr of the listening socket, so
+	 * we can use this information to connect to the listening socket.
+	 */
+	if (listenerSockaddr != NULL )
```
Comment (soumyadeep2007) [inline src/backend/cdb/motion/ic_udpifc.c:1339]: nit: trailing whitespace

**Review by soumyadeep2007 — APPROVED** [review]
LGTM! Minor comment added. Also, if you could just write a sentence in the commit message about the kinds of conflict you resolved.

I usually follow this template for backports:

```
This commit backports <>.
Conflicts resolved:
<...>
Original commit message follows:
<original commit message>
```

File: src/backend/cdb/motion/ic_udpifc.c (hunk 2 above)
Comment (Chibin) [reply src/backend/cdb/motion/ic_udpifc.c:1339]: Good catch! I thought I got rid of this :(

**Pull request merged**